src/dali/mapped_graph.py

```diff
@@ -0,0 +1,189 @@
+# Copyright 2023 Neal Chambers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import datetime, timedelta
+from typing import Dict, Iterator, List, NamedTuple, Optional, Set
+
+from prezzemolo.graph import Graph
+from prezzemolo.utility import ValueType
+from prezzemolo.vertex import Vertex
+from rp2.rp2_decimal import RP2Decimal
+from rp2.rp2_error import RP2TypeError
+
+from dali.historical_bar import HistoricalBar
+from dali.logger import LOGGER
+
+
+class Alias(NamedTuple):
+    from_asset: str
+    to_asset: str
+
+
+# Hard-coded base aliases
+_UNIVERSAL_ALIASES = {
+    Alias(from_asset="LUNA", to_asset="LUNC"): RP2Decimal("1"),
+    Alias(from_asset="XBT", to_asset="BTC"): RP2Decimal("1"),
+}
+
+_EXCHANGE_SPECIFIC_ALIASES = {
+    "Coinbase Pro": {
+        Alias(from_asset="ETH2", to_asset="ETH"): RP2Decimal("1"),
+    },
+    "Pionex": {
+        Alias(from_asset="MBTC", to_asset="BTC"): RP2Decimal("0.001"),
+        Alias(from_asset="METH", to_asset="ETH"): RP2Decimal("0.001"),
+    },
+}
+
+_UNIVERSAL: str = "UNIVERSAL"
+
+
+class MappedGraph(Graph[ValueType]):
+    def __init__(
+        self,
+        exchange: str,  # This is temporary until teleportation is implemented
+        vertexes: Optional[List["Vertex[ValueType]"]] = None,
+        optimized_assets: Optional[Set[str]] = None,
+        aliases: Optional[Dict[str, Dict[Alias, RP2Decimal]]] = None,
+    ) -> None:
+        super().__init__(vertexes)
+        self.__exchange: str = exchange  # Temporary until teleportation
+        self.__name_to_vertex: Dict[str, Vertex[ValueType]] = {vertex.name: vertex for vertex in vertexes} if vertexes else {}
+        self.__optimized_assets: Set[str] = set() if optimized_assets is None else optimized_assets
+        self.__universal_aliases: Dict[Alias, RP2Decimal] = {}
+        self.__universal_aliases.update(_UNIVERSAL_ALIASES)
+        self.__universal_aliases.update(_EXCHANGE_SPECIFIC_ALIASES.get(exchange, {}))  # To be removed for teleportation
+        # TO BE IMPLEMENTED - exchange specific aliases when teleportation is implemented
+        # self.__exchange_aliases: Dict[str, Dict[Alias, RP2Decimal]]
+
+        if aliases:
+            self.__universal_aliases.update(aliases.get(_UNIVERSAL, {}))
+            self.__universal_aliases.update(aliases.get(exchange, {}))
+        self.__add_aliases(self.__universal_aliases)
+
+    @property
+    def aliases(self) -> Iterator[Alias]:
```

Comment:
This property should have the same name as the underlying instance variable (e.g. universal_aliases).